crates/client/block_import/src/verify_apply.rs

```diff
@@ -0,0 +1,281 @@
+use std::{borrow::Cow, sync::Arc};
+
+use dc_db::{DeoxysBackend, DeoxysStorageError};
+use dp_block::{
+    header::PendingHeader, BlockId, BlockTag, DeoxysBlockInfo, DeoxysBlockInner, DeoxysMaybePendingBlock,
+    DeoxysMaybePendingBlockInfo, DeoxysPendingBlockInfo, Header,
+};
+use dp_convert::ToFelt;
+use starknet_api::core::ChainId;
+use starknet_core::types::Felt;
+use starknet_types_core::hash::{Poseidon, StarkHash};
+
+use crate::{
+    BlockImportError, BlockImportResult, PendingBlockImportResult, PreValidatedBlock, PreValidatedPendingBlock,
+    RayonPool, UnverifiedHeader, ValidatedCommitments, Validation,
+};
+
+mod classes;
+mod contracts;
+
+pub struct VerifyApply {
+    pool: Arc<RayonPool>,
+    backend: Arc<DeoxysBackend>,
+    // Only one thread at once can verify_apply.
+    mutex: tokio::sync::Mutex<()>,
+}
+
+impl VerifyApply {
+    pub fn new(backend: Arc<DeoxysBackend>, pool: Arc<RayonPool>) -> Self {
+        Self { pool, backend, mutex: Default::default() }
+    }
+
+    /// This function wraps the [`verify_apply_inner`] step, which runs on the rayon pool, in a tokio-friendly future.
+    pub async fn verify_apply(
+        &self,
+        block: PreValidatedBlock,
+        validation: Validation,
+    ) -> Result<BlockImportResult, BlockImportError> {
+        let _exclusive = self.mutex.lock();
+
+        let backend = Arc::clone(&self.backend);
+        self.pool.spawn_rayon_task(move || verify_apply_inner(&backend, block, validation)).await
+    }
+
+    /// See [`Self::verify_apply`].
+    pub async fn verify_apply_pending(
+        &self,
+        block: PreValidatedPendingBlock,
+        validation: Validation,
+    ) -> Result<PendingBlockImportResult, BlockImportError> {
+        let _exclusive = self.mutex.lock();
+
+        let backend = Arc::clone(&self.backend);
+        self.pool.spawn_rayon_task(move || verify_apply_pending_inner(&backend, block, validation)).await
+    }
+}
+
+/// This needs to be called sequentially, it will apply the state diff to the db, verify the state root and save the block.
+/// This runs on the [`rayon`] threadpool however as it uses parallelism inside.
+pub fn verify_apply_inner(
```

Comment:
thats true we should do that

Reply:
we can't, because the db needs the block hash and global_state_root, and we don't have the global_state_root until we have computed the trie
we could overlap some of the db writes inside dc-db, like, everything except the actual block header which we could store last once we have the global state root, but this would make the interface between dc-db and block-import a mess i think
Each of these steps are parallelized internally, so I don't think we would see much of a perf change at all

I am adding a comment pointing to this potential improvement so that we can potentially investigate it deeper but I'm pretty sure this is not worth it